external/src/main/java/com/redhat/cloud/policies/engine/actions/plugins/notification/PoliciesAction.java

```diff
@@ -0,0 +1,190 @@
+package com.redhat.cloud.policies.engine.actions.plugins.notification;
+
+import com.fasterxml.jackson.annotation.JsonFormat;
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.databind.JsonSerializer;
+import com.fasterxml.jackson.databind.PropertyNamingStrategy;
+import com.fasterxml.jackson.databind.SerializerProvider;
+import com.fasterxml.jackson.databind.annotation.JsonNaming;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import com.fasterxml.jackson.datatype.jsr310.ser.LocalDateTimeSerializer;
+
+import java.io.IOException;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class PoliciesAction {
+    private String accountId;
+    private LocalDateTime timestamp;
+    private Context context;
+    private Set<Event> events;
```

Comment:
> `Event` does not implement `equals` and `hashCode`, is that OK?
That's OK.

> If we receive the same event data twice (assuming that's possible...),
Not possible, each entry is a policy that triggered.

The reason I'm using a `Set` rather than a `List` is because I read that `ConcurrentSkipListMap#merge` is NOT guaranteed to be applied once atomically. Does that make sense?

Reply:
> The function is NOT guaranteed to be applied once atomically only if the value is not present.

If I understand correctly that sentence from the Javadoc, it means that the mapping function could be executed more than once and therefore it should not have any side-effect (it should be idempotent). I don't see any side-effects in the current function so I think we're safe whether we use a `List` or a `Set`.

Do we need a `Set`? I'm not sure but maybe I missed something. But if we do because it would avoid duplicates, then `equals` and `hashcode` should be implemented in `Event`. Otherwise, the `Set` won't do anything about duplicates.

Reply:
The default equals implementation the same as a `==` (checking if is the same reference), that works for this use case I think.

The current implementation adds to `existing` what we have in `addition`, if calling it twice with same `addition`, the set will be OK. With a `List` we will have duplicates (elements from addition twice in the list) which is what I'm trying to avoid here.
We don't need to implement equals/hashcode as we know that the data will come only once, so reference checking is fine in this case.

Reply:
Oh I was assuming that there could be different `Event` objects references for a single event. There's no problem if the reference is always the same so I'm resolving this.